repo2docker/buildpacks/conda/install-miniforge.bash

```diff
@@ -19,27 +18,27 @@ unset HOME
 wget --quiet $URL -O ${INSTALLER_PATH}
 chmod +x ${INSTALLER_PATH}
 
-# check md5 checksum
-if ! echo "${MD5SUM}  ${INSTALLER_PATH}" | md5sum  --quiet -c -; then
-    echo "md5sum mismatch for ${INSTALLER_PATH}, exiting!"
+# check sha256 checksum
+if ! echo "${SHA256SUM}  ${INSTALLER_PATH}" | sha256sum  --quiet -c -; then
+    echo "sha256 mismatch for ${INSTALLER_PATH}, exiting!"
     exit 1
 fi
 
 bash ${INSTALLER_PATH} -b -p ${CONDA_DIR}
 export PATH="${CONDA_DIR}/bin:$PATH"
 
-# Allow easy direct installs from conda forge
-conda config --system --add channels conda-forge
-
 # Do not attempt to auto update conda or dependencies
 conda config --system --set auto_update_conda false
 conda config --system --set show_channel_urls true
 
 # bug in conda 4.3.>15 prevents --set update_dependencies
 echo 'update_dependencies: false' >> ${CONDA_DIR}/.condarc
 
-# install conda itself
-if [[ "${CONDA_VERSION}" != "${MINICONDA_VERSION}" ]]; then
+# install conda itself, if needed
+# FIXME: MINIFORGE_VERSION might have a build number, like -1.
+# This will force conda to be installed each time
+# Use a better way of finding conda version!
+if [[ "${CONDA_VERSION}" != "${MINIFORGE_VERSION}" ]]; then
```

Comment:
from an image size perspective, it's probably good to do a `conda clean -yaf`after any solves to clean up the repodata and any cached or no-longer used tarballs and cache dirs... but the whole check is probably not necessary, as the convention is mini* is always released with the full conda version you're getting... if anything, you could check said version (and proper activation) with:

```bash
source ${CONDA_DIR}/bin/activate
conda --version | grep ${CONDA_VERSION}
```

to fail in the event of a bad upstream release.

Reply:
Good point on the `conda clean -yaf`. A little later (outside of this script) we run several more `conda` commands, [after which we run `clean -yaf`](https://github.com/jupyter/repo2docker/blob/26542dfc85c0a9567d2f266d51810c3a85244880/repo2docker/buildpacks/conda/__init__.py#L280). Is there a benefit to also running it earlier here? Naively it seems we trade-off image size (smaller if we also "clean" here) vs download time (clean later) because we might have to re-download things in the second call. But I am not sure if we'd ever have to re-download stuff in the second step... :-/ Does anyone know?

Reply:
well, if this layer precedes user stuff getting installed, it would get cached, no? with `update_conda` turned off, you're unlikely to change these based packages and therefore wouldn't download anything... of course, if someone grabs a new python, all bets are off...

Reply:
So the conclusion is: need to try out "keep using same Python version but install a few random packages" as well as "switch to a different version of Python and a few random packages" and see which of the options "clean up early" and "clean up later" produce a smaller image and if there are (significant) time differences in terms of network traffic.

Does `conda clean -afy` also remove the repodata? That file is somehow in my head as "big file you need to download before you can solve and that takes a lot of time when peopel think 'the solver is slow'".

Reply:
yes, `yaf` is the whole deal. You can get mid-grained performance info from, e.g.

```bash
conda clean -yaf 
CONDA_INSTRUMENTATION_ENABLED=1 conda create -vv --dry-run -n test_env -c conda-forge python pandas notebook
```

```
INFO conda.common.io:log_totals(635): TOTAL     0.000        13 conda.core.prefix_data.PrefixData.load
INFO conda.common.io:log_totals(635): TOTAL    13.842         1 conda.core.solve.Solver._collect_all_metadata
INFO conda.common.io:log_totals(635): TOTAL     0.000         1 conda.core.solve.Solver._find_inconsistent_packages
INFO conda.common.io:log_totals(635): TOTAL     1.496         1 conda.core.solve.Solver._run_sat
INFO conda.common.io:log_totals(635): TOTAL     0.744         2 conda.resolve.Resolve.gen_clauses
INFO conda.common.io:log_totals(635): TOTAL     1.021         3 conda.resolve.Resolve.get_reduced_index
INFO conda.common.io:log_totals(635): TOTAL     1.069         1 conda.resolve.Resolve.solve
```

That just created 51mb of json and pickled state... and didn't even install anything!

Just re-running the part after clean (with the warm cache) shows:

```bash
INFO conda.common.io:log_totals(635): TOTAL     0.000        13 conda.core.prefix_data.PrefixData.load
INFO conda.common.io:log_totals(635): TOTAL     1.057         1 conda.core.solve.Solver._collect_all_metadata
INFO conda.common.io:log_totals(635): TOTAL     0.000         1 conda.core.solve.Solver._find_inconsistent_packages
INFO conda.common.io:log_totals(635): TOTAL     1.553         1 conda.core.solve.Solver._run_sat
INFO conda.common.io:log_totals(635): TOTAL     0.828         2 conda.resolve.Resolve.gen_clauses
INFO conda.common.io:log_totals(635): TOTAL     0.988         3 conda.resolve.Resolve.get_reduced_index
INFO conda.common.io:log_totals(635): TOTAL     1.139         1 conda.resolve.Resolve.solve
```

Thing is, conda-forge probably statistically adds a new package every couple minutes, and it compares by repodata etag, so you're probably going to end up downloading the repodata again by the end of a long build across multiple layers.

My thinking is, new releases of conda (and therefore mini*) happen on a stately, just-about-quarterly pace, and as such this layer should be pretty darned stable. i'm always somewhat :crying_cat_face:  when i see a binder build cache-miss on it.

Reply:
We had this because miniconda was lagging behind conda versions, so we had to upgrade conda to get newer versions.

Reply:
I also want to keep this PR to be as close to a simple drop-in replacement for miniconda as possible though.

Reply:
Sounds like we should run `conda clean -yaf` here.

For more caching of this layer also join in over at https://github.com/jupyter/repo2docker/issues/854. I reckon right now we miss the cache on this layer (even with miniconda) because we install apt-get packages before we come to this layer. So repos that have things to apt-get ~always get a cache miss.

Reply:
Just realized we already do run it here, in https://github.com/jupyter/repo2docker/blob/26542dfc85c0a9567d2f266d51810c3a85244880/repo2docker/buildpacks/conda/install-miniconda.bash#L79. It happens after all the conda related stuff happens, but since it's in the script it's in the same layer. That's all we need, right?

Reply:
Or I could be confused, and the conda clean command needs to be moved somewhere else? But as long as it is in this script, it should be ok right?

Reply:
Yes, as long as it is in the script it will be in the same layer. I don't think it needs moving inside the script. (I am sure I looked for that line in this script earlier ... but missed it :-/)

I think this resolves this conversation. If you agree, click the resolve button :)